[WIP] Moving files and exclude files and paths

codingHahn opened the pull request:
Needs #4

codingHahn committed:
Initial string-based exclude mechanism

The implementation takes every path in a given path and checks if it is
mentioned in the exclude vector. If that isn't the case, push it onto a
new vector which only contains non-excluded paths. At the end, print all
non-excluded files.

INFO: Waites on the parser implementation and the split between
`exclude_strings` and `exclude_regex`

codingHahn committed:
Implement moving files to backup location

The current implementations can handle files and directories. If a file
already exists in the backup folder, it checks the last modified time to
determine if it should be overridden.

The implementations is now split into two files: path_filter which
concerns itself with filtering paths, as the name suggests. The other
file is named backup, where the main logic of the backup process happens
(comparing timestamps, moving files, moving directories recursively)

LucRome commented on src/configuration.rs:
```diff
@@ -1,3 +1,4 @@
+// TODO: Make the ConfStruct a global singleton
```
Great idea! I'll create an Issue so we can discuss it.

LucRome commented on src/main.rs:
```diff
@@ -1,5 +1,29 @@
+    // Backup every found file
+    for path in path_wo_excludes {
+        backup::backup(&path, &dest_folder).unwrap();
+    }
```
Maybe put these in a seperate Function? (Altho I don't know if it's worth the hustle)

LucRome commented on src/path_filter.rs:
```diff
@@ -0,0 +1,21 @@
+                let p = path.file_name().into_string().unwrap();
+                if !conf.exclude_patterns.contains(&p) {
+                    result.push(path.path());
+                }
```
Also check the Regex (merge the main Branch into this Branch to to use them), otherwise it seems perfect

LucRome commented:
Nice, it looks really good and it also seems to be workin

codingHahn committed:
Integrate new CLI-interface and make more robust

The detection, if the dest is included in the source should now be more
robust. In addition to that, the application now handles missing
permission to the destination folder more gracefully.

codingHahn committed:
Prelimanry regex exclude support

codingHahn committed:
Cleanup of control flow and code quality adjustments

main is now cleaned up and the first effort to modularise backup has
been made. The paths will now get filtered in `backup` instead of `main`
which will ensure that even nested paths respect the excluded
paths/regexes. This _should_ also handle symlinks, but this is not
tested.

codingHahn commented:
We need to figure out how to test this function. It is reliant on filesystem state, which is a problem in static unit tests. We could write integration tests in something like python or go the painful way: mocking the rust io API
